File: lessons/dsa/complexity-as-a-budget/examples/measure-scans.ts
/**
 * Count the steps two duplicate checks spend on the same input.
 *
 * The point of the lesson is that the step count, not the wall clock, is what
 * grows with the input, so both functions return the comparisons they made.
 */

/** Compare every item with every later item. Cost grows with n squared. */
export function stepsForPairwiseScan(values: readonly number[]): number {
  let steps = 0;
  for (let left = 0; left < values.length; left += 1) {
    for (let right = left + 1; right < values.length; right += 1) {
      steps += 1;
      if (values[left] === values[right]) return steps;
    }
  }
  return steps;
}

/** Read each item once against a set of what was already seen. */
export function stepsForSingleScan(values: readonly number[]): number {
  let steps = 0;
  const seen = new Set<number>();
  for (const value of values) {
    steps += 1;
    if (seen.has(value)) return steps;
    seen.add(value);
  }
  return steps;
}

File: lessons/dsa/complexity-as-a-budget/examples/measure_scans.cpp
// Count the steps two duplicate checks spend on the same input.
//
// The point of the lesson is that the step count, not the wall clock, is what
// grows with the input, so both functions return the comparisons they made.
#include <unordered_set>
#include <vector>

// Compare every item with every later item. Cost grows with n squared.
int stepsForPairwiseScan(const std::vector<int>& values) {
  int steps = 0;
  for (std::size_t left = 0; left < values.size(); ++left) {
    for (std::size_t right = left + 1; right < values.size(); ++right) {
      ++steps;
      if (values[left] == values[right]) {
        return steps;
      }
    }
  }
  return steps;
}

// Read each item once against a set of what was already seen.
int stepsForSingleScan(const std::vector<int>& values) {
  int steps = 0;
  std::unordered_set<int> seen;
  for (int value : values) {
    ++steps;
    if (!seen.insert(value).second) {
      return steps;
    }
  }
  return steps;
}
